### ISSUE TITLE
GamePad mappings are wrong (help wanted)

### ISSUE THREAD

**Issue by mrhelmut** [opentk/opentk#333]
Hello,

We recently discovered ([mono/MonoGame#4252](https://github.com/mono/MonoGame/issues/4252)) that most ```GamePad mappings``` in [https://github.com/opentk/opentk/blob/develop/Source/OpenTK/Input/GamePadConfigurationDatabase.cs](https://github.com/opentk/opentk/blob/develop/Source/OpenTK/Input/GamePadConfigurationDatabase.cs) are wrong (e.g. inverted axis, inverted buttons, etc.).

Fixing this issue will require to update the mapping parsing to allow a syntax to specify when axis shall be inverted.

If you're fine with this approach, I'll try submitting two separate PR: one to allow inverted axis, and one to update the mappings.

I would also like to request some help to test the mappings. If you are encountering mapping issues, it would be very useful to share the following:
- most importantly: the gamepad GUID (```Joystick.GetGuid(int index)```)
- the operating system
- the list of wrong buttons/axis and their apparent mapping destination

Gamepad priority are Xbox 360 (standard and wireless), Xbox One, Steam Controller, PS3 (standard and DualShock 3) & PS4 controllers.

It is possible that gamepads have different mappings based on how they are plugged in (e.g. PS4 controllers don't have the same mapping / Guid whether they are connected via bluetooth or USB), please test all configurations if possible.

EDIT: it may also be possible that mappings are different depending on the driver (SDL or GLControl), would that be possible that the raw values are different?

EDIT 2: confirmed, some GLControl mappings look totally fine, while SDL ones are wrong.

**Comment by endimilojkoski** [comment]
This is huge problem for me!
When I run my game through Steam client the gamepad controls are all messed up!

Only Left Stick is working fine on xbox 360 controller.
Big Button is button A.
Left Button is start.
Right Button is back.
Left Stick Button is Left Button;
Right Stick Button is Right Button.
Start is Right Stick Button.
Back is Left Stick Button;
Button A is DPad Up.
Button B is DPad Down.

I hope this is resolved ASAP or someone tells me how to unlink SDL2.dll from my game when I run it through Steam client!
Raining Blobs is scheduled for release on 15th of this month and I must resolve this somehow.
Any help is much appreciated!

**Comment by mrhelmut** [comment]
Wow, that's very messed up. Providing the Giud would help a lot. I have very different values on a 360 controller, I believe it's a different model.

I know that there are at least 2 xbox 360 controller revisions, there's a high chance that they are not wired the same and have different Giud.

**Comment by endimilojkoski** [comment]
I just built OpenTK from develop branch and the GUID is: 78696e70-7574-0000-0000-000000000000
I am testing with wired xbox 360 controller.

When I add SDL2.dll to the game the GUID is: 78696e70-7574-0100-0000-000000000000

I got the GUID with OpenTK.Input.Joystick.GetGuid(0)

Could the difference in GUID be the issue here???

**Comment by endimilojkoski** [comment]
Yes, the GUID difference is the issue!
Check in the Database file there is XInput record for first GUID and there is nothing for the second GUID when SDL2 is present!

What are the options now?

**Comment by mrhelmut** [comment]
Ahhh! So it's the Guid which are wrong, and possibly not the mapping!

One of the driver returns wrong GUID, not sure which one, investigating. Thanks!

(and that could possibly be related to other bugs where SDL thinks keyboards are gamepads)

**Comment by endimilojkoski** [comment]
I am glad I could help! Please let me know ASAP if you find anything.

**Comment by endimilojkoski** [comment]
Meanwhile I found this: http://www.opentk.com/node/3685
With this I can avoid SDL2.dll while this is fixed!
Thanks to @cra0zy for the link!

**Comment by mrhelmut** [comment]
Good to know! Be aware that GamePad will not work without SDL on OS X.

**Comment by mrhelmut** [comment]
Update.

There is no need to rework the mapping parsing. I noticed that Y's are inverted on all gamepad brands, hence I submitted #334.

Regarding the mapping, what I tested so far:
- Xbox 360 wired controller: OK on all platforms except Windows SDL (because of the GUID problem)
- PS4 controller (bluetooth): OK on all platforms
- PS4 controller (USB): no mapping for Windows, but OK on all other platforms

To be tested:
- standard PS3 controller (USB & bluetooth)
- DualShock 3 controller (USB & bluetooth)

Missing controller I'll probably add later:
- Steam controller
